File: VB/WpfApplication1/WizardCustomization/CustomSelectDataSourcePage.vb
#Region "#SelectDataSourcePage"
Imports Microsoft.VisualBasic
Imports DevExpress.Data.Utils.ServiceModel
Imports DevExpress.Data.XtraReports.DataProviders
Imports DevExpress.Data.XtraReports.ServiceModel
Imports DevExpress.Data.XtraReports.Wizard
Imports DevExpress.Data.XtraReports.Wizard.Presenters
Imports DevExpress.Data.XtraReports.Wizard.Views
Imports System
Imports System.Collections.Generic
Imports System.Linq
' ...

Namespace WpfApplication1.WizardCustomization
	Friend Class CustomSelectDataSourcePage
		Inherits SelectDataSourcePage(Of ReportModel)
		Private datasources As IEnumerable(Of DataSourceInfo)
		Public Sub New(ByVal view As ISelectDataSourcePageView, ByVal client As IReportWizardServiceClient)
			MyBase.New(view, client)
			AddHandler client.GetDataSourcesCompleted, AddressOf client_GetDataSourcesCompleted
		End Sub

		Private Sub client_GetDataSourcesCompleted(ByVal sender As Object, ByVal e As ScalarOperationCompletedEventArgs(Of IEnumerable(Of DataSourceInfo)))
			' Obtain the available data sources.	
			RemoveHandler Me.Client.GetDataSourcesCompleted, AddressOf client_GetDataSourcesCompleted
			datasources = e.Result
		End Sub

		Public Overrides Function GetNextPageType() As Type
			' The type of the next wizard page is chosen based on the type of the selected data source.
			Dim info = datasources.First(Function(x) x.Name = View.SelectedDataSourceName)
			Return If(info.TablesOrViewsSupported, GetType(CustomSelectDataMemberPage), GetType(CustomSelectHierarchicalDataSourceColumnsPage))
		End Function
	End Class
End Namespace
#End Region ' #SelectDataSourcePage

File: CS/WpfApplication1/WizardCustomization/CustomSelectDataSourcePage.cs
#region #SelectDataSourcePage
using DevExpress.Data.Utils.ServiceModel;
using DevExpress.Data.XtraReports.DataProviders;
using DevExpress.Data.XtraReports.ServiceModel;
using DevExpress.Data.XtraReports.Wizard;
using DevExpress.Data.XtraReports.Wizard.Presenters;
using DevExpress.Data.XtraReports.Wizard.Views;
using System;
using System.Collections.Generic;
using System.Linq;
// ...

namespace WpfApplication1.WizardCustomization {
    class CustomSelectDataSourcePage : SelectDataSourcePage<ReportModel> {
        IEnumerable<DataSourceInfo> datasources;
        public CustomSelectDataSourcePage(
            ISelectDataSourcePageView view, 
            IReportWizardServiceClient client)
            : base(view, client) {
            client.GetDataSourcesCompleted += client_GetDataSourcesCompleted;
        }

        void client_GetDataSourcesCompleted(object sender, ScalarOperationCompletedEventArgs<IEnumerable<DataSourceInfo>> e) {
            // Obtain the available data sources.
            this.Client.GetDataSourcesCompleted -= client_GetDataSourcesCompleted;
            datasources = e.Result;
        }

        public override Type GetNextPageType() {
            // The type of the next wizard page is chosen based on the type of the selected data source.
            var info = datasources.First(x => x.Name == View.SelectedDataSourceName);
            return info.TablesOrViewsSupported ? typeof(CustomSelectDataMemberPage) : typeof(CustomSelectHierarchicalDataSourceColumnsPage);
        }
    }
}
#endregion #SelectDataSourcePage